core/src/query.rs

```diff
@@ -30,7 +36,116 @@ pub struct MongoQuery {
+}
+
+impl Translation {
+    pub fn from_document(doc: &Document) -> Result<Self> {
+        let as_bson = Bson::Document(doc.clone());
+        let deserializer = bson::Deserializer::new(as_bson);
+        let deserializer = serde_stacker::Deserializer::new(deserializer);
+        Deserialize::deserialize(deserializer).map_err(Error::QueryDeserialization)
+    }
+}
+
+impl Namespace {
+    pub fn from_bson(bson: Bson) -> Result<BTreeSet<Self>> {
+        let deserializer = bson::Deserializer::new(bson);
+        let deserializer = serde_stacker::Deserializer::new(deserializer);
+        Deserialize::deserialize(deserializer).map_err(Error::QueryDeserialization)
+    }
+}
+
 impl MongoQuery {
+    fn get_sql_query_namespaces(sql_query: &str, db: &String) -> Result<BTreeSet<Namespace>> {
+        let get_namespaces_command = doc! {
+            "command": "getNamespaces",
+            "options": {
+                "sql": sql_query,
+                "db": db,
+            },
+        };
+
+        let returned_doc = handle_libmongosqltranslate_command(get_namespaces_command)?;
+
+        let namespaces: BTreeSet<Namespace> = Namespace::from_bson(
+            returned_doc
+                .get("namespaces")
+                .expect("The `namespace` key is missing from the document returned by libmongosqltranslate.")
+                .to_owned(),
+        )?;
+
+        Ok(namespaces)
+    }
+
+    fn translate_sql(
+        sql_query: &str,
+        current_db: &String,
+        namespaces: BTreeSet<Namespace>,
+        client: &MongoConnection,
+        db: &Database,
+    ) -> Result<Translation> {
+        let schema_collection = db.collection::<Document>("__sql_schemas_");
+
+        // create the schema_catalog document
+        let mut db_doc = doc! {};
+
+        for namespace in namespaces {
+            let namespace_schema_doc = client
+                .runtime
+                .block_on(async {
+                    schema_collection
+                        .find_one(doc! {
+                            "_id": &namespace.collection
+                        })
+                        .await
+                        .map_err(Error::QueryExecutionFailed)
+                })?
+                .ok_or(Error::SchemaDocumentNotFoundInSchemaCollection(
+                    namespace.collection.clone(),
+                ))?;
+
+            let bson_schema = namespace_schema_doc
+                .get("schema")
+                .expect("The required `schema` field is missing in the found Document.");
+
+            db_doc.insert(namespace.collection, bson_schema);
+        }
+
+        let schema_catalog_doc: Document = doc! {
+            current_db: db_doc
+        };
+
+        let translate_command = doc! {
+            "command": "translate",
+            "options": {
+                "sql": sql_query,
+                "db": current_db,
+                "excludeNamespaces": false,
+                "relaxSchemaChecking": true,
```

Comment:
Is true the default for relax schema checking?  I was looking for where that is set, I'll have to update it in my PR to true.

Reply:
I couldn't find it used anywhere, and when I looked into `mongohouse`, it seemed like they ignore it. 

This is their TranslateSQL function, however, they don't even have a parameter for `relaxSchemaChecking` mode, so I just always chose the default value (relaxed).  

```
func TranslateSQL(dbName, sqlStatement string, schemas NamespaceToSchemaMap, excludeNamespaces bool) (Translation, error) {
	translation, err := mongosql_rs.Translate(mongosql_rs.TranslationArgs{
		DB:                dbName,
		SQL:               sqlStatement,
		CatalogSchema:     schemas,
		ExcludeNamespaces: excludeNamespaces,
	})
	if err != nil {
		return Translation{}, convertSafeSQLError(err)

	}

	ret := Translation{
		Pipeline:         translation.Pipeline,
		TargetDB:         translation.TargetDB,
		TargetCollection: translation.TargetCollection,
		ResultSetSchema:  translation.ResultSetSchema,
		SelectOrder:      translation.SelectOrder,
	}
	return ret, nil
}
```